packages/loader/container-loader/src/connectionHandler.ts

```diff
@@ -0,0 +1,889 @@
+} from "@fluidframework/driver-utils";
+import {
+    GenericError,
+} from "@fluidframework/container-utils";
+import { DeltaQueue } from "./deltaQueue";
+import {
+    ReconnectMode,
+    IConnectionManager,
+    IConnectionManagereFactoryArgs,
+} from "./contracts";
+
+const MaxReconnectDelayInMs = 8000;
+const InitialReconnectDelayInMs = 1000;
+const DefaultChunkSize = 16 * 1024;
+
+function getNackReconnectInfo(nackContent: INackContent) {
+    const message = `Nack (${nackContent.type}): ${nackContent.message}`;
+    const canRetry = nackContent.code !== 403;
+    const retryAfterMs = nackContent.retryAfter !== undefined ? nackContent.retryAfter * 1000 : undefined;
+    return createGenericNetworkError(
+        `nack [${nackContent.code}]`, message, canRetry, retryAfterMs, { statusCode: nackContent.code });
+}
+
+const createReconnectError = (fluidErrorCode: string, err: any) =>
+    wrapError(
+        err,
+        (errorMessage: string) => new GenericNetworkError(fluidErrorCode, errorMessage, true /* canRetry */),
+    );
+
+/**
+ * Implementation of IDocumentDeltaConnection that does not support submitting
+ * or receiving ops. Used in storage-only mode.
+ */
+class NoDeltaStream
+    extends TypedEventEmitter<IDocumentDeltaConnectionEvents>
+    implements IDocumentDeltaConnection, IDisposable
+{
+    clientId: string = "storage-only client";
+    claims: ITokenClaims = {
+        scopes: [ScopeType.DocRead],
+    } as any;
+    mode: ConnectionMode = "read";
+    existing: boolean = true;
+    maxMessageSize: number = 0;
+    version: string = "";
+    initialMessages: ISequencedDocumentMessage[] = [];
+    initialSignals: ISignalMessage[] = [];
+    initialClients: ISignalClient[] = [];
+    serviceConfiguration: IClientConfiguration = {
+        maxMessageSize: 0,
+        blockSize: 0,
+        summary: undefined as any,
+    };
+    checkpointSequenceNumber?: number | undefined = undefined;
+    submit(messages: IDocumentMessage[]): void {
+        this.emit("nack", this.clientId, messages.map((operation) => {
+            return {
+                operation,
+                content: { message: "Cannot submit with storage-only connection", code: 403 },
+            };
+        }));
+    }
+    submitSignal(message: any): void {
+        this.emit("nack", this.clientId, {
+            operation: message,
+            content: { message: "Cannot submit signal with storage-only connection", code: 403 },
+        });
+    }
+
+    private _disposed = false;
+    public get disposed() { return this._disposed; }
+    public dispose() { this._disposed = true; }
+}
+
+/**
+ * Implementation of IConnectionManager, used by Container class
+ * Implements constant connectivity to relay service, by reconnecting in case of loast connection or error.
+ * Exposes various controls to influecen this process, including manual reconnects, forced read-only mode, etc.
+ */
+export class ConnectionManager implements IConnectionManager {
```

Comment:
i'm not doing a super deep review of this class. it looks primarily like a copy paste, with minor changes to support encapsulation, which is awesome. On a superficial review it looks good.  if there is a particular area of concern here, like an area that changes a lot and you like a second pair of eyes on feel free to ping me in a comment there and i will look further.

Reply:
i would love to see unit test for this new class, but fine if they come later. is there is an issue for that?

Reply:
Not really :) I think API shape and the particular place where I made a cut is more interesting to discuss - it took me couple iterations to arrive to what I have, trying to keep minimal amount of knowledge that each layer has to know about another one (but also allow highest level - Container, to control unique characteristics that DM does not care about).

Reply:
The parts that changed (based on memory)
- order of things happenning on connection ("connect" event now fires later than it was, after initial ops are processed).
- The way we prepare op before sending (new concept, extracted from "send") changed slightly semantics, which caused ops to be accumulated in send buffer and exposed the fact that container runtime was flushing ops while in offline (addressed on both sides - DM will log an error and drop them, Container Runtime is fixed not to flush when in offline)

Reply:
@ChumpChief has a good comment about shape below. however, i'm not super concerned about shape right now. this seems like a good approach to disentangle the layers as minimally as possible. before we expose anything as an externally available abstraction we should go through design review.